fortran: do not link Fortran libs with unnecessary libraries

ggouaillardet opened the pull request:
Fortran libraries do not need to be linked with PMIx, hwloc nor libevent. That can cause some issues with incompatible flags are pulled from the .la files.
For example, when the NAG Fortran compiler is used, libmpi_mpifh_sizeof.la pulls the -pthread flag, that is translated into -Wl,-pthread.
Then when libmpi_mpifh.la is linked with gcc but pulls -Wl,-pthread that causes the build to fail since -pthread is not a valid ld option. Not depending on these libraries is enough to avoid the above issue.

Refs #12413

Thanks to Matthew Thompson for reporting this issue and helping with the diagnostic.

hppritcha commented:
@ggouaillardet would you have time to resolve the conflicts?  it would be nice to get this in to main and v6.0.x

ggouaillardet committed:
fortran: do not link Fortran libs with unnecessary libraries

Fortran libraries do not need to be linked with PMIx, hwloc nor
libevent. That can cause some issues with incompatible flags are pulled
from the .la files.
For example, when the NAG Fortran compiler is used,
libmpi_mpifh_sizeof.la pulls the -pthread flag, that is translated
into -Wl,-pthread.
Then when libmpi_mpifh.la is linked with gcc but pulls -Wl,-pthread that
causes the build to fail since -pthread is not a valid ld option.
Not depending on these libraries is enough to avoid the above issue.

Refs #12413

Thanks to Matthew Thompson for reporting this issue and helping with
the diagnostic.

Signed-off-by: Gilles Gouaillardet <gilles@rist.or.jp>

ggouaillardet committed:
test/asm: fix checks when cross compiling

When running the riscv CI, each test program must be run under QEMU.
This is done automatically if a wrapper is generated, but no wrapper
is issued since the test programs do not depend on any libtool generated
dynamic library any more.
This commit forces the test program to depend on libopen-pal.la,
and hence force a wrapper to be generated and have the tests
automatically invoked under QEMU

Signed-off-by: Gilles Gouaillardet <gilles@rist.or.jp>

ggouaillardet committed:
riscv64: run autogen.pl in a rockylinux 9 container

Signed-off-by: Gilles Gouaillardet <gilles@rist.or.jp>

ggouaillardet commented:
@jsquyres that was a wild one!

1. build failed. when linking with `libmpi.so`, the paths to `libhwloc.so` and friends is not passed to the link command line. I could not make any sens of this, so I tried to `autogen.pl` with `rockylinux9` and the latest `archlinux` and it worked (!)
2. the `test/asm` failed. since none of these tests depend on a shared `libtool` generated library (e.g. `*.la`), no wrappers was issued and a test (e.g. `atomic_barrier`) is invoked directly instead of via `qemu` and hence crashed.

so on top of the initial changes, I implemented two workarounds
1. build a `rockylinux9` based `singularity` container just enough to run `autogen.pl`
2. have the tests in `test/asm` depend on `libopen-pal.la` in order to have wrappers generated.

I will double check the first point, maybe I did something wrong after all and got lucky with an other `libtool`.
Note `rockiylinux9` has older and `archlinux` has newer autotools compared to `ubuntu 24.04`, so the discrepancy might come from some `debian` patches (!).

ggouaillardet commented:
since `libmpi.la` explicitly calls `hwloc`, `pmix` and `libevent`, maybe I missed to add them

```patch
diff --git a/ompi/Makefile.am b/ompi/Makefile.am
index c7610c2..4e44e2f 100644
--- a/ompi/Makefile.am
+++ b/ompi/Makefile.am
@@ -145,7 +145,9 @@ lib@OMPI_LIBMPI_NAME@_la_LIBADD = \
 
 lib@OMPI_LIBMPI_NAME@_la_LIBADD += \
         $(OMPI_TOP_BUILDDIR)/opal/lib@OPAL_LIB_NAME@.la \
-        $(opal_pmix_LIBS)
+        $(opal_pmix_LIBS) \
+        $(opal_hwloc_LIBS) \
+        $(opal_libevent_LIBS)
 lib@OMPI_LIBMPI_NAME@_la_DEPENDENCIES = $(lib@OMPI_LIBMPI_NAME@_la_LIBADD)
 lib@OMPI_LIBMPI_NAME@_la_LDFLAGS = \
        -version-info $(libmpi_so_version) \
```

That being said, `ompi_info` still fails to link when `autogen.pl` is invoked with `ubuntu 24.04` autotools.
(no `-L...` is passed for those libraries, and they cannot be found).
At this stage, I have no clue of what is going wrong.

hppritcha commented:
Should we keep this PR open?

hppritcha commented:
closing we can re-open if there's interest.
